File: Blind_75_LeetCode_Questions/c++/Dynamic_Programming/Word_Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) 
    {
        unordered_set<string> word_dict(wordDict.begin(), wordDict.end());
        int word_len = s.size();

        bool segment[word_len + 1];
        memset(segment, false, sizeof(segment));
        segment[0] = true;

        for(int i = 1; i <= word_len; ++i)
        {
            for(int j = 0; j < i; ++j) 
            {
                if(segment[j] && word_dict.count(s.substr(j, i - j)))
                {
                    segment[i] = true;
                    break;
                }
            }
        }
        return segment[word_len];
        

    }
};

File: Blind_75_LeetCode_Questions/java/Dynamic_Programming/Q139_Word_Break/Word_Break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) 
    {
        int len = s.length();
        boolean[] tmp = new boolean[len + 1];
        tmp[0] = true; 
        
        for(int i = 1; i <= len; ++i)
        {
            for(int j = 0; j < i; ++j)
            {
                if(tmp[j] && wordDict.contains(s.substring(j, i)))
                {
                    tmp[i] = true;
                    break;
                }
            }
        }
        return tmp[len];
    }
}
